### ISSUE TITLE
Keep track of field usage in java

### ISSUE THREAD

**Issue by WindLeeWT** [protocolbuffers/protobuf#12615]
Protobuf is heavily used in our business and one of the most important message **Log** is recording all the online events and parameters for a single request in the system(composed of hundreds of micro services). Then **Log** messages are sent into message queue and consumed by lots of downstream application(backend services/big data applications) for monitoring/reporting/training.

After several years of development, there are large numbers of fields added into **Log** and **Log** become a very complicated message with some nested levels of sub-messages. The main pain point currently is that we don't know which fields are actually used by downstream application after compiled into java sources. The field usage is critical for us to do some field management, such as a field should be totally offlined if it was still in the message data but no one access it.

After some investigation, the most reliable way seems to add some hooks for ```getXXX``` methods and there is indeed some insertion points in C++
```c++
inline ::google::protobuf::int32 Log::test_field() const {
  // @@protoc_insertion_point(...)
  return test_field_;
}
inline void Log::set_test_field(::google::protobuf::int32 value) {
  set_has_test_field();
  test_field_ = value;
  // @@protoc_insertion_point(...)
}
```
However, I don't see these insertion points in java and I also saw a few issues/pull requests previously(#3292, #2684). My questions are:

1. Is there any other ways to do field usage collection and it'd be much helpful if there is some discussions/examples before
2. Whether hooks for ```getXXX``` is feasible or not, I knew java generated sources are much complicated than C++ as there are interfaces/builders/messages inside and I need to modify protoc plugin for my own hook code in Java

Thanks in advance

**Comment by fowles** [comment]
@WindLeeWT I recommend using https://kythe.io/ it provides cross references that are searchable and queryable that will let you do what you want.

**Comment by WindLeeWT** [comment]
> @WindLeeWT I recommend using https://kythe.io/ it provides cross references that are searchable and queryable that will let you do what you want.

Hi fowles, thanks for your reply and I took a close look at Kythe which is a language-agnostic protocols and data formats describing source code as well as a series of tools/utilities. 

However, we don't have a monolithic giant code repository like Google and we just built and packaged Java sources generated from protobuf, then released into a repository for downstream data applications in various of teams and repositories some of which we even don't know or have access. 

In other words, source code relationship/dependency is loose in our case and I'm trying to find possible ways to collect ```getXxx``` field usage at **runtime** given the common jar in repository is fully controlled by us, the maintainer of the ```Log```.

**Comment by fowles** [comment]
If you want to do *dynamic* collection of data (which it sounds like you do), I don't really have good answers.  I would look into using existing java dynamic code coverage tooling and try to reverse the mangling `protoc` uses.

**Comment by WindLeeWT** [comment]
> If you want to do _dynamic_ collection of data (which it sounds like you do), I don't really have good answers. I would look into using existing java dynamic code coverage tooling and try to reverse the mangling `protoc` uses.

Sorry for the late reply as we're on vacation and the issue was closed so quickly. Following the comment mentioned by @fowles above, I thought maybe there was some misunderstanding here. 

I was asking whether it's possible to insert **some code** into Java code generated by ```protoc``` and do some monitoring/reporting when the code is then executed. For instance, we're able to insert customized code for a specific field access in C++ with help of **insertion point** mechanism as below

```C++
/*
  optional int32 test_field = 1;
*/
inline ::google::protobuf::int32 Log::test_field() const {
  // @@protoc_insertion_point(...)
  if (the_first_time_) {
    // Report to a central service that test_field is **used** by someone in a machine(ip) at some point in time
    the_first_time_ = true;
  }
  return test_field_;
}
```

Of course the extra cost should be at the lowest level and that's in C++. I'd like to ask if we could do the same thing in Java(**insertion point or other mechanism**) or if there is other better ways to achieve the field usage collection.

@deannagarcia Please let the issue be open for a moment and wait for more discussion if someone else is interested, thanks

**Comment by fowles** [comment]
insertion points expose too many internal details and are extremely fragile, so we do not wish to add them to other languages.  You may be able to do something equivalent using bytecode weaving.  But we don't have any plans to add a feature like this to java protobuf.
